modules/infra/datapath/control_output.c

```diff
@@ -0,0 +1,65 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2024 Christophe Fontaine
+
+#include "gr_control_output.h"
+
+#include <gr_control.h>
+#include <gr_graph.h>
+#include <gr_log.h>
+#include <gr_mbuf.h>
+
+#include <rte_ether.h>
+#include <rte_graph_worker.h>
+
+#include <stdlib.h>
+
+enum {
+	NO_CALLBACK_ERROR,
+	ENQUEUE_ERROR,
+	EDGE_COUNT,
+};
+
+static uint16_t control_output_process(
+	struct rte_graph *graph,
+	struct rte_node *node,
+	void **objs,
+	uint16_t n_objs
+) {
+	struct gr_control_output_msg msg;
+	struct rte_mbuf *mbuf;
+
+	for (unsigned i = 0; i < n_objs; i++) {
+		mbuf = objs[i];
+		msg.callback = control_output_mbuf_data(mbuf)->callback;
+		msg.mbuf = mbuf;
+		if (msg.callback == NULL) {
+			rte_node_enqueue_x1(graph, node, NO_CALLBACK_ERROR, mbuf);
+			continue;
+		}
+
+		if (control_output_push(&msg) != 0) {
+			rte_node_enqueue_x1(graph, node, ENQUEUE_ERROR, mbuf);
+		}
+	}
+
+	if (n_objs)
+		signal_control_ouput_message();
+
+	return n_objs;
+}
+
+static struct rte_node_register control_output_node = {
+	.name = "control_output",
+	.process = control_output_process,
+	.nb_edges = EDGE_COUNT,
+	.next_nodes = {[NO_CALLBACK_ERROR] = "ctrl_out_no_callback", [ENQUEUE_ERROR] = "ctrl_out_enqueue"},
```

Comment:
not really important, but can you move to one edge per line, like the other nodes?